Mathlib/Tactic/Linter/CommandStart.lean

```diff
@@ -0,0 +1,149 @@
+as well as all `variable` commands.
+-/
+def lintUpTo (stx : Syntax) : Option String.Pos :=
+  if let some cmd := stx.find? (·.isOfKind ``Lean.Parser.Command.declaration) then
+    match cmd.find? (·.isOfKind ``Lean.Parser.Term.typeSpec) with
+      | some s => s.getPos?
+      | none => match cmd.find? (·.isOfKind ``Lean.Parser.Command.declValSimple) with
+        | some s => s.getPos?
+        | none => none
+  else if stx.isOfKind ``Lean.Parser.Command.variable then
+    stx.getTailPos?
+  else none
+
+def removeComments (s : String) : String :=
+  let lines := s.splitOn "\n"
+  let lines := lines.filterMap fun l =>
+    -- remove lines that begin with a comment
+    if (l.trim.startsWith "--") || (l.trim.startsWith "/-" && l.trim.get ⟨2⟩ != '-') then none
+    -- remove the text in a line, starting from the beginning `--`
+    else if let st::_ := l.splitOn "--" then
+      -- FIXME! make sure that we do not truncate a doc-string!
+      --if st.back == '/' then some l else some
+       st.trimLeft
+    else some l
+  "\n".intercalate lines
+/-
+#eval do
+  let s := "Hi\n  -- this is a comment\nthere there is some -- another comment -- with more\ntext"
+  IO.println <| removeComments s
+-/
+
+def furtherFormatting (s : String) : String :=
+  s |>.replace "¬ " "¬"
+    |>.replace "aesop (rule_sets" "aesop(rule_sets"
+    |>.replace " Prop." " «Prop»."
+    |>.replace " Type." " «Type»."
+
+namespace Style.CommandStart
+
+/--
+`unlintedNodes` contains the `SyntaxNodeKind`s for which there is no clear formatting preference:
+if they appear in surface syntax, the linter will ignore formatting.
+
+Currently, the unlined nodes are mostly related to `Subtype`, `Set` and `Finset` notation and
+list notation.
+-/
+abbrev unlintedNodes := #[``«term_::_», ``«term{_:_//_}», `«term{_}», `Mathlib.Meta.setBuilder,
+  `Bundle.termπ__, `Finset.«term_#_»]
+
+@[inherit_doc Mathlib.Linter.linter.style.commandStart]
+def commandStartLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  unless Linter.getLinterValue linter.style.commandStart (← getOptions) do
+    return
+  if (← get).messages.hasErrors then
+    return
+  -- If a command does not start on the first column, emit a warning.
+  if let some pos := stx.getPos? then
+    let colStart := ((← getFileMap).toPosition pos).column
+    if colStart ≠ 0 then
+      Linter.logLint linter.style.commandStart stx
+        m!"'{stx}' starts on column {colStart}, \
+          but all commands should start at the beginning of the line."
+  -- We only lint up to the position given by `lintUpTo`
+  if let some finalLintPos := lintUpTo stx then
+    if let some stype := stx.find? (unlintedNodes.contains ·.getKind) then
+      Linter.logLintIf linter.style.commandStart.verbose (stx.getHead?.getD stx)
+        m!"Found '{stype.getKind}' in '{stype}'"
+      if let some pos := stype.getPos? then
+        if pos.1 ≤ finalLintPos.1 then
+          return
+    let stx := capSyntax stx finalLintPos.1
```

Comment:
I didn't scrutinise the tests file yet: it would be great to test this line, to clarify the behaviour. (Basically, my ideal is that if the tests pass, a linter change is correct: if mathlib exposes further edge cases, adding them as tests would be great.) With that in mind, does the file already have tests for
- unlinted nodes being left alone
- syntax before an unlinted node gets linted?